File: src/lib/tools.ts
import { invoke } from "@tauri-apps/api/core";

export interface FfmpegStatus {
  installed: boolean;
  downloading: boolean;
  /** 0–100 enquanto baixa. */
  progress: number;
  error: string | null;
}

export async function ffmpegStatus(): Promise<FfmpegStatus> {
  return invoke("ffmpeg_status");
}

/** Baixa o ffmpeg em segundo plano (volta na hora); acompanhe via `ffmpegStatus`. */
export async function ffmpegInstall(): Promise<void> {
  return invoke("ffmpeg_install");
}

/** Tradução automática com cache no banco; devolve o original se falhar. */
export async function translateText(text: string, target: string): Promise<string> {
  return invoke("translate_text", { text, target });
}

File: src-tauri/src/commands/tools.rs
use crate::{ffmpeg, state::AppState};
use tauri::{AppHandle, Manager};

/// Estado do ffmpeg baixado sob demanda (Config mostra "baixando 42%" etc.).
#[tauri::command]
pub fn ffmpeg_status(app: AppHandle) -> ffmpeg::InstallStatus {
    ffmpeg::status(&app)
}

/// Dispara o download do ffmpeg em segundo plano (volta na hora) e, quando
/// terminar, já roda o pós-processamento pendente — chamado quando o
/// usuário liga uma das opções, pra não esperar o próximo ciclo de 30min.
#[tauri::command]
pub fn ffmpeg_install(app: AppHandle) {
    tauri::async_runtime::spawn(async move {
        let http = app.state::<AppState>().http.clone();
        if ffmpeg::ensure_installed(&app, &http).await.is_ok() {
            crate::postprocess::spawn_pending(&app, None);
        }
    });
}

/// Sinopse traduzida pro idioma `target` ("pt"). Cai no texto original se a
/// tradução falhar (sem internet, limite do serviço etc.).
#[tauri::command]
pub async fn translate_text(
    state: tauri::State<'_, AppState>,
    text: String,
    target: String,
) -> Result<String, ()> {
    Ok(crate::sources::translate::translate(&state.http, &state.db, &text, &target).await)
}
